lib/services/nativescript-migration-service.ts

```diff
@@ -179,5 +304,95 @@ export class NativeScriptMigrationService implements IFrameworkMigrationService
+		return path.join(this.tnsModulesDirectoryPath, NativeScriptMigrationService.TYPESCRIPT_ABBREVIATION, `${version}${NativeScriptMigrationService.TNS_CORE_MODULES}.zip`);
+	}
+
+	private getBackupName(str: string): string {
+		return `${str}.backup`;
+	}
+
+	private promptForConfirmation(message: string): IFuture<void> {
+		return (() => {
+			let shouldMigrate = this.$prompter.confirm(message, () => true).wait();
+			if (!shouldMigrate) {
+				this.$errors.failWithoutHelp('Operation canceled.');
+			}
+		}).future<void>()();
+	}
+
+	private extractTnsModulesPackage(newVersion: string, tnsModulesProjectPath: string): IFuture<void> {
+		return (() => {
+			let projectType = this.$project.isTypeScriptProject().wait() ? NativeScriptMigrationService.TYPESCRIPT_ABBREVIATION : NativeScriptMigrationService.JAVASCRIPT_ABBREVIATION;
+			let pathToNewTnsModules = path.join(this.tnsModulesDirectoryPath, projectType, this.getFileNameByVersion(newVersion));
+			this.$fs.unzip(pathToNewTnsModules, tnsModulesProjectPath).wait();
+			this.moveAppResourcesDirectory(this.nativeScriptMigrationConfiguration.appResourcesObsoletePath, this.nativeScriptMigrationConfiguration.appResourcesRequiredPath).wait();
+		}).future<void>()();
+	}
+
+	private moveAppResourcesDirectory(from: string, to: string): IFuture<void> {
+		return (() => {
+			if (this.$fs.exists(from).wait() && !this.$fs.exists(to).wait()) {
+				this.$logger.trace(`Moving ${from} to ${to}`);
+				this.$fs.rename(from, to).wait();
+			}
+		}).future<void>()();
+	}
+
+	private traceError(err: Error): void {
+		this.$logger.trace("Error during migration. Trying to restore previous state.");
+		this.$logger.trace(err);
+	}
+
+	private deleteOriginalTnsModulesDirectory(): IFuture<void> {
+		return this.$fs.deleteDirectory(this.nativeScriptMigrationConfiguration.tnsModulesProjectPath);
+	}
+
+	private restoreOriginalTnsModulesDirectory(): IFuture<void> {
+		return (() => {
+			this.deleteOriginalTnsModulesDirectory().wait();
+			this.$fs.rename(this.nativeScriptMigrationConfiguration.tnsModulesBackupName, this.nativeScriptMigrationConfiguration.tnsModulesProjectPath).wait();
+		}).future<void>()();
+	}
+
+	private executeMigrationActions(jsonContents:any, oldPath: string, newPath: string, beforeRenameAction: () => IFuture<void>, afterRenameAction: () => IFuture<void>): IFuture<void> {
+		return (() => {
+			this.$fs.writeJson(this.nativeScriptMigrationConfiguration.pathToPackageJson, jsonContents).wait();
+			if (this.$project.isTypeScriptProject().wait()) {
+				beforeRenameAction().wait();
+
+				if (this.$fs.exists(oldPath).wait()) {
+					this.$fs.rename(oldPath, newPath).wait();
+				}
+
+				afterRenameAction().wait();
+			}
+		}).future<void>()();
+	}
+
+	private getModuleVersion(version: string): IFuture<string> {
+		return (() => {
+			return _.find(this.nativeScriptMigrationData.wait().supportedVersions, sv => sv.version === version).modulesVersion;
```

Comment:
this code will fail in case the find returns undefined:
```JavaScript
var a = [{ v: '1', m: '2' }, { v: '3', m: '4'}];
var l = require("lodash");
> l.find(a, function(el) { return el.v === '3'}).m
'4'
> l.find(a, function(el) { return el.v === '6'}).m
TypeError: Cannot read property 'm' of undefined
    at repl:1:47
    at REPLServer.defaultEval (repl.js:132:27)
    at bound (domain.js:254:14)
    at REPLServer.runBound [as eval] (domain.js:267:12)
    at REPLServer.<anonymous> (repl.js:279:12)
    at REPLServer.emit (events.js:107:17)
    at REPLServer.Interface._onLine (readline.js:214:10)
    at REPLServer.Interface._line (readline.js:553:8)
    at REPLServer.Interface._ttyWrite (readline.js:830:14)
    at ReadStream.onkeypress (readline.js:109:10)
```